content/blog/2023/09/03/2023-09-03-jsr305-probe-blog.adoc

```diff
@@ -0,0 +1,51 @@
+---
+layout: post
+title: Detecting deprecated JSR-305 imports Probe
+tags:
+- gsoc
+- gsoc2023
+- healthscore
+- probes
+- plugin
+authors:
+- jagruti
+opengraph:
+  image: /images/gsoc/2023/jagruti/jsr-305-probe-image.png
+sig:
+- gsoc
+---
+
+== Introduction
+
+The `JSR-305` framework provides annotations to detect software defects. Its executive committee voted it as dormant since May 2012 (see: link:https://jcp.org/en/jsr/detail?id=305[JSR 305: Annotations for Software Defect Detection
+])
+
+It is important to stop using the framework for the following reasons:
+
+- It is not maintained anymore.
+- The annotations have questionable license.
+- The annotations are in the reserved javax namespace and there is no public release of the spec (Nor is there ever likely to be. See point 1).
+(Source: link:https://groups.google.com/g/jenkinsci-dev/c/uE1wwtVi1W0/m/gLxdEJmlBQAJ[Jenkins Developers Discussion])
+
+
+`@Nonnull` and `@CheckForNull` annotations (part of JSR-305) have been link:https://www.jenkins.io/doc/developer/tutorial-improve/replace-jsr-305-annotations/[deprecated since 2016].
+
+The `JSR-305` probe checks whether the deprecated annotations are used in a plugin.
+
+== Importance of the probe
+
+This probe will identify the outdated annotations used in the plugin and highlight them in the scoring system.
+
+== Challenges and Learning
+
+The outdated annotations are imported in `.java` files. My first approach was to go for an abstract class because I assumed there could be other probes that may require to look for code in `.java` files.
+After a series of reviews, one of the mentors pointed out that to comply with the link:https://github.com/openrewrite/rewrite-jenkins/blob/98bfbecc9c9507a069c9ac1f1e35cd5dd0fa0687/src/main/resources/META-INF/rewrite/rewrite.yml[OpenWrite rule] all imports from the `javax.annotation` package should be checked for deprecation.
+
+Also, based on the link:https://blog.devgenius.io/key-principles-in-software-dry-kiss-yagni-solid-and-other-acronyms-98e5575a6942[YAGNI principle] ("You Aren't Gonna Need It") which means develop features only when you need it, the `abstract` wasn't the current requirement. I restructured the code and used a single class to implement the probe.
```

Comment:
```suggestion
== Challenges and Learning

The outdated annotations are imported in `.java` files.
My first approach was to go for an abstract class because I assumed there could be other probes that may need to look for code in `.java` files.
After a series of reviews, one of the mentors pointed out that to comply with the link:https://github.com/openrewrite/rewrite-jenkins/blob/98bfbecc9c9507a069c9ac1f1e35cd5dd0fa0687/src/main/resources/META-INF/rewrite/rewrite.yml[OpenWrite rule], all imports from the `javax.annotation` package should be checked for deprecation.

Also, based on the link:https://blog.devgenius.io/key-principles-in-software-dry-kiss-yagni-solid-and-other-acronyms-98e5575a6942[YAGNI principle] ("You Aren't Gonna Need It"), which means features are developed only when you need them, the `abstract` wasn't the current requirement.
I restructured the code and used a single class to implement the probe.
```

suggestion for formatting and some syntax

Were there any additional challenges in restructuring the code or using a single class?

Reply:
There weren't any other challenges. I am just sharing the thought process of arriving at the solution here to help the future GSoCers/contributors.